### DISCUSSION TITLE
Mesh stored as a linked-list or array

### DISCUSSION THREAD

**Discussion by vitorvas** [seamplex/feenox#22] [Ideas]
Hello Germán,

Any reason to store mesh data in a linked list instead of an array?
Just wondering if feenox tampers too much with the mesh or if it is mostly kept untouched after data is read. (talking about a gmsh mesh here).

Thank you!

**Comment by jeremyonscale** [comment]
Awesome question. Some things are stored in linked lists and some are stored in arrays.
The rule is: if I already know the total number of items, use an array. Otherwise, use a list.
Also, lists are used in "small" sets and arrays in "big" sets.
So the nodes and elements are stored as arrays. The physical groups and materials are stored in lists.

But everything can be improved!

I was always entertaining the idea of going back to data-oriented organization, i.e. instead of having an array of structs (say for the element), having several arrays of element ids, element types, connectivities, etc.

Sorry for the generality of the answer but the question was also general?
Maybe if you elaborate I can give you better directions.

**Reply by gtheler** [reply]
Sorry about the username, I was logged in with a different account and did not realize.

**Reply by vitorvas** [reply]
> Sorry for the generality of the answer but the question was also general? Maybe if you elaborate I can give you better directions.

Thanks for the fast reply.
Actually, I didn't go deep into the code yet. I was performing a profiling with a "fat" `bunny.msh` with 2,5 million elements and was thinking about improving the mesh reading process which turns to be the big time consumer.

![output](https://github.com/user-attachments/assets/4fcbceb1-3da7-4e8a-8482-0a4474a191c9)

**Reply by gtheler** [reply]
Oh, I now see what you mean. Indeed, there is a method `add_element_to_list()` that is used as follows.

Each node has a list of elements it belongs to. When the elements are parsed, for each node that make up the element being parsed, it is added to the node's list o elements: https://github.com/seamplex/feenox/blob/main/src/mesh/gmsh.c#L664

The rationale is that when one needs to evaluate a function defined over nodes in a mesh at an arbitrary point $(x,y,z)$, first a kd-tree search is performed to find the nearest node to $(x,y,z)$. And then, to find which element contains the point $(x,y,z)$, FeenoX loops over the elements the nearest node belongs to and checks which element contains $(x,y,z)$. Then it finds the coordinates of $(x,y,z)$ in the element's local system so it can then compute the shape functions used to weight the nodal values.

This kd-tree and list of elements a node belongs to is only needed if one needs to evaluate funcions defined over meshes. This can be a direct evaluation like $T(0.1,0.2,0.3)$ or a non-conformal mapping from one mesh to another (say you read $T(x,y,z)$ from one mesh and use it to evaluate material properties on another mesh).

The first thing we can do here is to add an option to disable the tree and element lists.
The second thing would be to try to detect after parsing the input file if these are needed.

Can you share the mesh? I have the feeling that the `READ_MESH` instruction is I/O bounded anyway.

**Reply by vitorvas** [reply]
Thanks for clarifying. It is always easier to learn from the author.
Yes, I can share the mesh. It is yours, just refined.

The gzipped version of the file is 51MB, larger than allowed here. Do you have any suggestion on how to share it?

**Reply by gtheler** [reply]
ok, can you paste the geo you used to get it?

**Reply by vitorvas** [reply]
I didn't use it. I just refined the .msh file you already provide on your feenox test set using `gmsh` itself. Tomorrow I can check the details, but I think I just used the `-3` key with gmsh one or two times over the original.`bunny.msh`

**Reply by gtheler** [reply]
Ok, so I changed the max element size from 15 down to 0.8 and I got a 2.5 million element count.
Indeed, FeenoX needs 15-ish seconds to read it and Gmsh only 3 seconds:

```
gtheler@tom:~/codigos/feenox/examples$ time ../feenox read_mesh.fee bunny-big.msh 
2.55268e+06
417753

real    0m16.273s
user    0m15.839s
sys     0m0.408s
gtheler@tom:~/codigos/feenox/examples$ time gmsh - bunny-big.msh 
Info    : Running 'gmsh - bunny-big.msh' [Gmsh 4.13.0-git-022b171c7, 1 node, max. 1 thread]
Info    : Started on Thu Aug 29 10:24:55 2024
Info    : Reading 'bunny-big.msh'...
Info    : 55 entities
Info    : 417753 nodes
Info    : 2552684 elements                                              
Info    : 18 parametrizations                                              
Info    : [  0%] Discrete surface 2 is planar, simplifying parametrization
Info    : [ 20%] Discrete surface 6 is planar, simplifying parametrization            
Info    : [ 30%] Discrete surface 7 is planar, simplifying parametrization            
Info    : [ 40%] Discrete surface 10 is planar, simplifying parametrization           
Info    : [ 60%] Discrete surface 13 is planar, simplifying parametrization           
Info    : [ 70%] Discrete surface 14 is planar, simplifying parametrization           
Info    : [ 70%] Discrete surface 15 is planar, simplifying parametrization
Info    : [ 80%] Discrete surface 16 is planar, simplifying parametrization           
Info    : [ 80%] Discrete surface 17 is planar, simplifying parametrization
Info    : [ 90%] Discrete surface 18 is planar, simplifying parametrization           
Info    : [ 90%] Discrete surface 19 is planar, simplifying parametrization
Info    : Done reading 'bunny-big.msh'ations                                          
Info    : Stopped on Thu Aug 29 10:24:58 2024 (From start: Wall 2.92098s, CPU 2.81037s)

real    0m3.017s
user    0m2.877s
sys     0m0.140s
```

When using binary meshes the difference is even larger because Gmsh does optimize the binary reading but FeenoX does not (it reads one value at a time):

```
gtheler@tom:~/codigos/feenox/examples$ time ../feenox read_mesh.fee bunny-big-binary.msh 
2.55268e+06
417753

real    0m15.421s
user    0m14.881s
sys     0m0.372s
gtheler@tom:~/codigos/feenox/examples$ time gmsh - bunny-big-binary.msh 
Info    : Running 'gmsh - bunny-big-binary.msh' [Gmsh 4.13.0-git-022b171c7, 1 node, max. 1 thread]
Info    : Started on Thu Aug 29 10:25:24 2024
Info    : Reading 'bunny-big-binary.msh'...
Info    : 55 entities
Info    : 417753 nodes
Info    : 2552684 elements                                              
Info    : 18 parametrizations                                              
Info    : [  0%] Discrete surface 2 is planar, simplifying parametrization
Info    : [ 20%] Discrete surface 6 is planar, simplifying parametrization            
Info    : [ 30%] Discrete surface 7 is planar, simplifying parametrization            
Info    : [ 40%] Discrete surface 10 is planar, simplifying parametrization           
Info    : [ 60%] Discrete surface 13 is planar, simplifying parametrization           
Info    : [ 70%] Discrete surface 14 is planar, simplifying parametrization           
Info    : [ 70%] Discrete surface 15 is planar, simplifying parametrization
Info    : [ 80%] Discrete surface 16 is planar, simplifying parametrization           
Info    : [ 80%] Discrete surface 17 is planar, simplifying parametrization
Info    : [ 90%] Discrete surface 18 is planar, simplifying parametrization           
Info    : [ 90%] Discrete surface 19 is planar, simplifying parametrization
Info    : Done reading 'bunny-big-binary.msh'                                         
Info    : Stopped on Thu Aug 29 10:25:24 2024 (From start: Wall 0.561995s, CPU 0.448293s)

real    0m0.647s
user    0m0.519s
sys     0m0.128s
```


Now, commenting out the call to `feenox_mesh_add_element_to_list()` in https://github.com/seamplex/feenox/blob/main/src/mesh/gmsh.c#L776 we now perform better than Gmsh:

```
gtheler@tom:~/codigos/feenox/examples$ time ../feenox read_mesh.fee bunny-big.msh 
2.55268e+06
417753

real    0m1.814s
user    0m1.664s
sys     0m0.148s
```

Maybe someone with a PhD in computer science can come up with a good idea. Either an alternative to detecting which element an arbitrary space location belongs to that does not depend on these element lists or a way of computing those lists only if needed.

**Reply by vitorvas** [reply]
You got **way** better times than me. I will repeat my tests and share the whole data.
It also must be said that the computer I used to run it is below average.
I won't be able to do it today, I will check tomorrow.

Thanks!

**Reply by vitorvas** [reply]
Germán, by the way, could you share here the `read_mesh.fee` file? If we are comparing things I want to do it as close as you do. Thanks.

**Reply by gtheler** [reply]
Yes, sorry about that.

```
gtheler@tom:~/codigos/feenox/examples$ cat read_mesh.fee 
READ_MESH $1
PRINT elements
PRINT nodes
gtheler@tom:~/codigos/feenox/examples$ cat bunny.geo 
// https://upload.wikimedia.org/wikipedia/commons/4/43/Stanford_Bunny.stl
Merge "Stanford_Bunny.stl";
angle = 40;
includeBoundary = 1;
forceParametrizablePatches = 0;
curveAngle = 180;

ClassifySurfaces{angle * Pi/180, includeBoundary, forceParametrizablePatches, curveAngle * Pi / 180};
CreateGeometry;

// Create a volume as usual
Surface Loop(1) = Surface{:};
Volume(1) = {1};

Physical Volume("bunny") = {1};
Physical Surface("rest") = {3};
Physical Surface("base") = {2};

Mesh.MeshSizeMin = 0.8;
Mesh.MeshSizeMax = 0.8;
Mesh.ElementOrder = 1;
Mesh.MeshSizeExtendFromBoundary = 0;

gtheler@tom:~/codigos/feenox/examples$ 
```

**Reply by vitorvas** [reply]
```
vitors@DT32528CDTN-LX:~/workspace/feenox$ time feenox read_mesh.fee bunny.msh 
2.46555e+06
408867

real	0m14,735s
user	0m13,608s
sys	0m0,816s
```

It is not your big bunny, but mine.
Yet, no considerable difference between your test and mine. That's not what I expected.

I will dig into my notes and come back with the test I did. (What annoys me is that I only find time to "play" with feenox once here and there and I'm always coming back to remember what I was doing the last time...)

Regards,

Vitor

**Reply by vitorvas** [reply]
![read_mesh](https://github.com/user-attachments/assets/5545dfdf-cf16-445d-bde4-04f3900d9448)


Just for completeness, the profile of `read_mesh.fee`

**Reply by gtheler** [reply]
Yes, one needs a couple of hours to enter into [flow state](https://www.medicalnewstoday.com/articles/flow-state#flow-vs-hyperfocus).

Maybe your mesh is finer than mine. Feel free to comment out the call that adds elements to the node's lists and re-profile.

**Comment by gtheler** [comment]
Hey @vitorvas . Where are we on this subject?

**Comment by vitorvas** [comment]
Hello German,

Didn't touch it since your explanation.
Unfortunately, I have no plans at this time. Sorry.

Regards,

Vitor

Em sáb., 28 de dez. de 2024 às 11:22, jeremy theler <
***@***.***> escreveu:

**Reply by gtheler** [reply]
I was thinking that maybe replacing a list with a dynamic array or even hash would speed things up.
Then I realized that I was appending to a list each time a new element arrived, while I could directly prepend since I do not care about the order in the list: https://github.com/seamplex/feenox/commit/b3b4312d141e133fae9bfdced98883a25cba9d35

Theoretically, that should change $O(n)$ into $O(1)$.

@vitorvas can you run your test again and see if that profile improves and that 79% goes down?

**Comment by vitorvas** [comment]
Hi Germán,

Sorry for the delay.
Well, the machine I was using to test it is gone. I had it replaced and I
do not even have a proper Linux on the new one. Actually, there are no
Linux options anymore to work with.
I will check if I still have access to the data on the old machine (if the
machine is still physically on the premises) Sorry about that.
Unfortunately, I work in a research center where the IT department works as
if we were an ordinary business. It is a nightmare.

I will let you know if I get it.

Regards,

Vitor

Em qua., 12 de mar. de 2025 às 12:19, jeremy theler <
***@***.***> escreveu:

**Reply by gtheler** [reply]
Don't worry. I'll profile with cachegrind before and after the fix.
I'll post the results here.
